### PR TITLE
Issue-38 Add Alcohol By Weight calculator

### PR DESCRIPTION
**Pull request by ProgrammerJoe93**
Added ABV to ABW calculator for #38

### CODE REVIEW

File: rustybeer-cli/src/commands/alcohol_volume_weight.rs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+                .help("'From' alcohol percentage")
+                .required(true)
+                .takes_value(true),
+        )
+        .arg(
+            Arg::with_name("total_volume")
+                .short("v")
+                .long("total_volume")
+                .value_name("TOTAL_VOLUME")
+                .help("Total beer volume in mL")
```
Comment (drodil) [inline rustybeer-cli/src/commands/alcohol_volume_weight.rs:26]: You could use VolumeBuilder for this to take any volume and then just convert to milliliters

File: rustybeer-cli/src/commands/alcohol_volume_weight.rs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+                .required(false)
+                .takes_value(true),
+        )
+        .arg(
+            Arg::with_name("total_density")
+                .short("d")
+                .long("total_density")
+                .value_name("TOTAL_DENSITY")
+                .help("Total density of beer in kg/m³")
```
Comment (drodil) [inline rustybeer-cli/src/commands/alcohol_volume_weight.rs:35]: I think there is density unit also in measurements project. I started working on the from_str functionalities there but didn't complete all of them so not sure if I did this. Other option is to create yet another builder for density

File: rustybeer-cli/src/commands/alcohol_volume_weight.rs (hunk 1 above)
Comment (ProgrammerJoe93) [reply rustybeer-cli/src/commands/alcohol_volume_weight.rs]: Added

File: rustybeer-cli/src/commands/alcohol_volume_weight.rs (hunk 2 above)
Comment (ProgrammerJoe93) [reply rustybeer-cli/src/commands/alcohol_volume_weight.rs]: There seems to only be mass, volume and temperature. I'll see if I can add density to the conversion file in a different ticket.

File: README.md [hunk 3]
```diff
@@ -61,6 +61,7 @@ Implemented                   | Function      | Description
 :white_check_mark:            | Priming       | Beer Priming Calculator                        | `priming --temp <Beer temperature> --amount <Beer volume> --co2_volumes <co2_volumes>`
 :white_check_mark:            | SG Correction | Corrects SG reading according to the difference between the measurement temperature and the calibration temperature | `sg_correction --sg <Specific gravity reading> --ct <Calibration temperature> --mt <Measurement temperature>`
 :white_check_mark:            | Beer style    | Finds beer styles matching given parameters    | `beer_style (--og <Original gravity>) (--fg <Final gravity>) (--abv <Alcohol by volume>) (--ibu <International bittering units> (--color <SRM color>)`
+:white_check_mark:            | ABV <-> ABW    | Calculates alcohol by weight (ABW) from  alcohol by volume (ABV) | `abv_abw --percent <alcohol percentage> (--total_volume <total beer volume in ml>) (--total_density <density of beer in g/cm³) (--reverse)`
```
Comment (drodil) [inline README.md:64]: ml still here :)

File: rustybeer-cli/src/commands/alcohol_volume_weight.rs [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+};
+use rustybeer_util::conversions::VolumeBuilder;
+
+pub fn add_subcommand<'a, 'b>() -> App<'a, 'b> {
+    App::new("abv_abw")
+        .version("0.1")
+        .author("Joseph Russell (josephrussell123@gmail.com)")
+        .about("Calculates Alcohol by Weight (ABW) from Alcohol By Volue (ABV) and vice versa")
+        .arg(
+            Arg::with_name("percent")
```
Comment (drodil) [inline rustybeer-cli/src/commands/alcohol_volume_weight.rs:14]: Maybe abv is more used term in the tool for this?

File: rustybeer-cli/src/commands/alcohol_volume_weight.rs [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+                .value_name("PERCENT")
+                .help("'From' alcohol percentage")
+                .required(true)
+                .takes_value(true),
+        )
+        .arg(
+            Arg::with_name("total_volume")
+                .short("v")
+                .long("total_volume")
+                .value_name("TOTAL_VOLUME")
```
Comment (drodil) [inline rustybeer-cli/src/commands/alcohol_volume_weight.rs:26]: You could remove "total" and just use volume

File: rustybeer-cli/src/commands/alcohol_volume_weight.rs [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+                .help("Total beer volume")
+                .required(false)
+                .takes_value(true),
+        )
+        .arg(
+            Arg::with_name("total_density")
+                .short("d")
+                .long("total_density")
```
Comment (drodil) [inline rustybeer-cli/src/commands/alcohol_volume_weight.rs:34]: Same here, density should be good enough

File: rustybeer-cli/src/commands/alcohol_volume_weight.rs [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    if let Some(matches) = matches.subcommand_matches("abv_abw") {
+        if matches.is_present("percent") {
+            let percent = value_t!(matches, "percent", f64).unwrap_or_else(|e| e.exit());
+            let end_percentage: f64;
+
+            // main ABV <-> ABW conversion
+            if matches.is_present("reverse") {
+                if matches.is_present("total_density") {
+                    let total_density =
+                        value_t!(matches, "total_density", f64).unwrap_or_else(|e| e.exit());
```
Comment (drodil) [inline rustybeer-cli/src/commands/alcohol_volume_weight.rs:61]: As said earlier, you could create DensityBuilder to convert from string. But that can be done later or when the measurements library supports FromStr trait

File: rustybeer/src/calculators/alcohol_volume_weight.rs [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+pub fn calculate_abv_abw_density(percent: f64, total_density: f64) -> f64 {
+    percent * (ETHANOL_DENSITY / total_density)
+}
+
+pub fn calculate_abw_abv(percent: f64) -> f64 {
+    percent * 1.25
+}
+
+pub fn calculate_abw_abv_density(percent: f64, total_density: f64) -> f64 {
```
Comment (drodil) [inline rustybeer/src/calculators/alcohol_volume_weight.rs:24]: Tests missing but can be done in separate PR for the love of Hacktoberfest :D

File: README.md (hunk 3 above)
Comment (ProgrammerJoe93) [reply README.md]: fixed

File: rustybeer-cli/src/commands/alcohol_volume_weight.rs (hunk 5 above)
Comment (ProgrammerJoe93) [reply rustybeer-cli/src/commands/alcohol_volume_weight.rs]: fixed

File: rustybeer-cli/src/commands/alcohol_volume_weight.rs (hunk 6 above)
Comment (ProgrammerJoe93) [reply rustybeer-cli/src/commands/alcohol_volume_weight.rs]: fixed

**Review by drodil — APPROVED** [review]

**Pull request merged**